File: mesh-control-python-server/utils/edit_prov_db.py
import os
import json

def add_appkey(key, bound):
        try:
                if not key:
                        return "Cannot add an empty key!"
                elif not bound:
                        return "Cannot add a app key without bounded netkey!"
                
                home_path = os.path.expanduser('~')
                meshctl_path = home_path + "/.config/meshctl/"
                f = open(meshctl_path+"prov_db.json","r")
                mesh_info = json.loads(f.read())
                mesh_info["appKeys"].append({
                "index":len(mesh_info["appKeys"]),
                "boundNetKey":int(bound),
                "key":key
                })
                f.close()
                f = open(meshctl_path+"prov_db.json","w")
                f.write(json.dumps(mesh_info))
                f.close()
                
                return f"Succesfully added a new key: {key}"
        except Exception as e:
                return f"Failed: {e}"
        
def edit_appkey(keyData):
        try:
                if not keyData["key"]:
                        return "Cannot edit an empty key!"
                
                home_path = os.path.expanduser('~')
                meshctl_path = home_path + "/.config/meshctl/"
                f = open(meshctl_path+"prov_db.json","r")
                mesh_info = json.loads(f.read())
                edit_index = None
                for i in range(0,len(mesh_info["appKeys"])):
                        print(mesh_info["appKeys"][i]["index"],keyData["index"])
                        if mesh_info["appKeys"][i]["index"] == int(keyData["index"]):
                                edit_index = i
                print(edit_index)
                mesh_info["appKeys"][edit_index] = { 
                "index":int(keyData["index"]),
                "boundNetKey":int(keyData["boundNetKey"]),
                "key":keyData["key"]
                }
                f.close()
                f = open(meshctl_path+"prov_db.json","w")
                f.write(json.dumps(mesh_info))
                f.close()
                
                return f"Succesfully edited a key {edit_index}: {keyData['key']}"
        except Exception as e:
                return f"Failed: {e}"
        
def remove_appkey(index):
        try:
                home_path = os.path.expanduser('~')
                meshctl_path = home_path + "/.config/meshctl/"
                f = open(meshctl_path+"prov_db.json","r")
                mesh_info = json.loads(f.read())

                removed_key = mesh_info["appKeys"][index]
                del mesh_info["appKeys"][index]
                
                f.close()
                f = open(meshctl_path+"prov_db.json","w")
                f.write(json.dumps(mesh_info))
                f.close()
                
                return f"Succesfully removed a key: {removed_key['key']}"
        except Exception as e:
                return f"Failed: {e}"
        

File: mesh-control-client/pages/api/utils/editProvdb.ts
import fs from 'fs';
import os from 'os';

import {NodesConfig, AppKey} from '../../../interfaces/global'

const homePath = os.homedir()
const meshctlProvdb = `${homePath}/.config/meshctl/prov_db.json`

export function addAppkey(key: string, bound: string | number): string {
  try {
    if (!key) {
      return "Cannot add an empty key!";
    }
    if (!bound) {
      return "Cannot add a app key without bounded netkey!";
    }

    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);
    
    meshInfo.appKeys.push({
      index: meshInfo.appKeys.length,
      boundNetKey: parseInt(bound.toString()),
      key: key
    });
    
    fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
    
    return `Successfully added a new key: ${key}`;
  } catch (e) {
    return `Failed: ${e}`;
  }
}

export function editAppkey(keyData: AppKey): string {
  try {
    if (!keyData.key) {
      return "Cannot edit an empty key!";
    }

    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);
    
    let editIndex: number | null = null;
    for (let i = 0; i < meshInfo.appKeys.length; i++) {
      console.log(meshInfo.appKeys[i].index, keyData.index);
      if (meshInfo.appKeys[i].index === parseInt(keyData.index.toString())) {
        editIndex = i;
        break;
      }
    }
    
    if (editIndex === null) {
      return "Key index not found!";
    }
    
    console.log(editIndex);
    meshInfo.appKeys[editIndex] = {
      index: parseInt(keyData.index.toString()),
      boundNetKey: parseInt(keyData.boundNetKey.toString()),
      key: keyData.key
    };
    
    fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
    
    return `Successfully edited a key ${editIndex}: ${keyData.key}`;
  } catch (e) {
    return `Failed: ${e}`;
  }
}

export function removeAppkey(index: number): string {
  try {
    
    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);

    if (index >= meshInfo.appKeys.length || index < 0) {
      return "Invalid key index!";
    }

    meshInfo.appKeys.splice(index, 1);
    global.DATA.CONFIG.NODES.appKeys = meshInfo.appKeys;
    global.DATA.KEYS.APPKEYS = meshInfo.appKeys;
    fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
    
    return `Successfully removed a key: ${index}`;
  } catch (e) {
    return `Failed: ${e}`;
  }
}

export function removeNode(address: string) {
  try {
    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);
   for (let i = 0; i < meshInfo.nodes.length; i++) {
    for (let j = 0; j < meshInfo.nodes[i].configuration.elements.length; j++) {
      if (meshInfo.nodes[i].configuration.elements[j].unicastAddress === address) {
        meshInfo.nodes.splice(i, 1);
        global.DATA.CONFIG.NODES.nodes = meshInfo.nodes;
        fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
        return "Node removed!";
      }
    }
   }
   return "Node not found!";
  } catch (e) {
    return `Failed: ${e}`;
  }
}

export function resetNodesList() {
  try {
    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);
    meshInfo.nodes = [];
    fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
  } catch (e) {
    return `Failed: ${e}`;
  }
}

export function resetAppkeysList() {
  try {
    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);
    meshInfo.appKeys = [];
    fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
  } catch (e) {
    return `Failed: ${e}`;
  }
}

export function resetNetkeysList() {
  try {
    const fileContent = fs.readFileSync(meshctlProvdb, 'utf8');
    const meshInfo: NodesConfig = JSON.parse(fileContent);
    meshInfo.netKeys = [];
    fs.writeFileSync(meshctlProvdb, JSON.stringify(meshInfo));
  } catch (e) {
    return `Failed: ${e}`;
  }
}